### ISSUE TITLE
`TSInstall` and `TSInstallSync` fail when plugin directory readonly

### ISSUE THREAD

**Issue by cideM** [nvim-treesitter/nvim-treesitter#436]
**Describe the bug**
I installed `nvim-treesitter` as per the documentation, but trying to `TSInstall` gives me

> `Error executing vim.schedule lua callback: vim.lua:254: Expected lua string`

and `TSInstallSync`

```
Downloading...
Compiling...
E5108: Error executing lua ...obar/opt/nvim-treesitter/lua/nvim-treesitter/install.lua:52: Expected lua string
```

**Expected behavior**
No error

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

    health#nvim_treesitter#check
    ========================================================================
    ## Installation
      - OK: `git` executable found.
      - OK: `cc` executable found.

</code>
</details>

**Output of `nvim --version`**
```
NVIM v0.5.0-dev
Build type: Release
LuaJIT 2.1.0-beta3
Compilation:
Compiled by nixbld

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/nix/store/62w9q8wc0yapiffgfz8zg311zxcsli93-neovim-unwrapped-master/share/nvim
"

Run :checkhealth for more info

```

**Additional context**

*I'm using Nix to install neovim*. Just in case `nvim-treesitter` is trying to write something to the Neovim directory or a path relative to Neovim install or relative to the `init.vim` location (which are all in the immutable Nix store on my machine) then this operation would fail.

I tried this with various `nvim-treesitter` commits but they all cause the same issue which apparently no one else has.

**Comment by theHamsta** [comment]
Hi. Could replace the cited line:
```lua
      api.nvim_err_writeln(cmd.err) 
```
to
```lua
  api.nvim_err_writeln(vim.inspect(cmd))
```
This should give us more context for the error. Either moving the compiled parser to it's suposed location or delete the compilation directory failed (we should really make sure to print a meaningful error when the `cmd` has no `err` field.

**Comment by theHamsta** [comment]
Hi. Could replace the cited line: ter/lua/nvim-treesitter/install.lua:52
```lua
      api.nvim_err_writeln(cmd.err) 
```
to
```lua
  api.nvim_err_writeln(vim.inspect(cmd))
```
This should give us more context for the error. Either moving the compiled parser to it's suposed location or delete the compilation directory failed (we should really make sure to print a meaningful error when the `cmd` has no `err` field.

**Comment by cideM** [comment]
Adding some info for the next person who reads this. I manage Neovim plugins through Nix and manage sources with `niv`. Here's how I modified `nvim-treesitter` do add the debugging statement `vim.inspect`

```
  nvim-treesitter = (pkgs.vimUtils.buildVimPluginFrom2Nix rec {
    name = "nvim-treesitter";
    src = vimPluginsSources.nvim-treesitter;
    postInstall = ''
      sed -i -e 's/api.nvim_err_writeln(cmd.err)/api.nvim_err_writeln(vim.inspect(cmd))/' $out/share/vim-plugins/nvim-treesitter/lua/nvim-treesitter/install.lua
    '';
  });
```

and here's the output of running `TSInstallSync`

```
Downloading...
Compiling...
{
  cmd = "mv",
  opts = {
    args = { "/home/cloud/.local/share/nvim/tree-sitter-lua/parser.so", "/nix/store/j5yjh93h2y638rkh71ml8vl01ph7546a-vim-pack-dir/pack/foobar/opt/nvim-treesitter/parser/lua.so" }
  }
}
```

I guess this is trying to move the file into `/nix/store` which won't work since that is read-only.

**Comment by theHamsta** [comment]
We're checking the compile and cache directory for for RW access, but always assume that the plugin directory is always readable. We need some directory in Vim's runtime path to move the parsers to. Do you have any suggestion to which directory we could fall back in case we can't write to the plugin directory?

**Comment by cideM** [comment]
Anything in `/nix/store` won't work. I guess a standard `XDG_DATA_DIR` as you suggested would work. But, now that I'm thinking about this a bit more I realize that installing the parsers through the tree sitter plugin is not actually a very Nix-y thing to do. Ideally Nix would handle the installation of these things. It shouldn't be too hard for me to download and compile the parsers through nix (just needs `git` and a C compiler after all). 

Do I have to add the compiled parsers to `nvim-treesitter/parser/*.so`? Since Nix itself can obviously write to the `/nix/store` it would be easy enough to do.

**Comment by theHamsta** [comment]
Any rtpath with subdirectory `parser` will do it. The parser location is a Neovim concept and not specific to our plugin. Before nvim-treesitter existed I just installed the parser directories as plugins and compiled the parser to a `parser` subdir of the repo.

**Comment by theHamsta** [comment]
You could also run `nvim --headless -c "TSInstallSync all" -c "q"` during the installation of this plugin. The directory should be writable during installation...

**Comment by cideM** [comment]
I now have 

```shell
$ ls /nix/store/pkh4d9szp4jcdizrhckg9yjbls1naq37-vim-pack-dir/pack/foobar/start/tree-sitter-go/parser/parser.so
```

and `/nix/store/pkh4d9szp4jcdizrhckg9yjbls1naq37-vim-pack-dir/pack/foobar/start/tree-sitter-go` is in my `rtp`

but if I run `TSInstallInfo` then `go` is not marked as installed.

I'm building it with 

```
      cc -o parser/parser.so -I./src -shared -Os -lstdc++ src/parser.c
```
and then I just copy the whole `parser/parser.so` to the output folder which ends up as `pack/foobar/start/tree-sitter-go/parser/parser.so` as shown above

based on [your Go treesitter files which apparently only need `src/parser.c`](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/parsers.lua#L56-L62)

**Comment by theHamsta** [comment]
The name should be `parser/go.so`

**Comment by cideM** [comment]
:facepalm: works perfectly now.

I'll see how I can turn these insights into a Nix PR to maybe make these treesitter modules available for the NixOS VIM ecosystem.

With that being said, it's always super cool if "upstream" can accommodate Nix users if it doesn't cause too much friction. If having another fallback location that's part of `rtp` would be a viable option so even with a Nix installed `nvim-treesitter` things can be installed through the plugin, that would be :rocket: 

What do you think?

**Comment by theHamsta** [comment]
Do you want to create the PR? Since you're already testing it... Would be great to make this plugin more resilent to changes in different enviroments.

**Comment by theHamsta** [comment]
Maybe you can use the `stdpath` function that causes less cross-platform problems.

```
stdpath({what})					*stdpath()* *E6100*
		Returns |standard-path| locations of various default files and
		directories.

		{what}       Type    Description ~
		cache        String  Cache directory. Arbitrary temporary
		                     storage for plugins, etc.
		config       String  User configuration directory. The
		                     |init.vim| is stored here.
		config_dirs  List    Additional configuration directories.
		data         String  User data directory. The |shada-file|
		                     is stored here.
		data_dirs    List    Additional data directories.

		Example: >
			:echo stdpath("config")
```

**Comment by cideM** [comment]
Closed by #473
